### PR TITLE
Add SDK provider configuration tests, part 3

### PR DESCRIPTION
**Pull request by BBBmau**
By adding tests for the providerConfigure function we will know what behaviours we need to test for in the plugin framework version of that logic. We want to assert parity between SDK and plugin framework versions of the code.

This PR references that tests that are implemented in #7723

Coverage:

- [x] batching
 

- - [x] send_after

- - [x] enable_batching
- [x] request_timeout 
- [x] request_reason

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [x] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/develop/run-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none
```

### CODE REVIEW

**Comment by NickElliot** [comment]
Sending this over to @SarahFrench for review as it is following up on their earlier changes!

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/provider/provider_internal_test.go:701]: Could you please change the name to reflect the expected outcome from the invalid value?

There's definitely a balance to be struck between too-short and too-long and wordy, but I think the test name should reflect that an error is involved

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/provider/provider_internal_test.go:711]: We know that the default value is set elsewhere now, so you can remove the TODOs :)

Instead, could you add some short comments explaining why the test shows the default value is not present? This could be link to a GitHub issue, or permalinks to other lines of code. What do you think would be most helpful to someone reading this for the first time?

**Review by SarahFrench — COMMENTED** [review]
👋 Hey @BBBmau , thanks so much for doing this work - looking good so far.

This is just a first-pass review from me. I haven't looked at all of it in depth yet but wanted to submit some review comments to help unblock you a little! I'll re-review as soon as I can

**Review by SarahFrench — COMMENTED** [review]
Here are some more bits of feedback - all looking great so far and it's great we're identifying fields whose default values aren't handled at the schema-level.

As well as the individual comments I've left on your diffs, could you also add a commit where you delete this test I wrote ages ago: [TestHandleSDKDefaults_RequestReason](https://github.com/GoogleCloudPlatform/magic-modules/blob/8cd4a506f0ac4db7b07a8cce914449d34df6f20b/mmv1/third_party/terraform/transport/config_test.go#L410-L484). This is because your new tests make that test redundant.

_Context_: Back when I started increasing the test coverage for the provider config code I started adding tests too granularly. I then swapped from testing `func HandleSDKDefaults` to testing `func providerConfigure`, which calls HandleSDKDefaults internally, and I've been deleting those original tests on HandleSDKDefaults as they become redundant.

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (BBBmau) [reply mmv1/third_party/terraform/provider/provider_internal_test.go:889]: Good call, I definitely agree with adding `ExpectedSendAfterValue:      ""` just to check that the default value is still there

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (BBBmau) [reply mmv1/third_party/terraform/provider/provider_internal_test.go:902]: I added this case but found that I was getting a fail due to no erroring showing up when inputting `-10s`. This shouldn't be the case right? @SarahFrench 

I also ended up adding a case where the user sets the value as `10` where they forget to include the seconds `s`.

Comment (SarahFrench) [reply mmv1/third_party/terraform/provider/provider_internal_test.go:902]: Ah, I think the validation isn't applied during the course of our test, that is the code we're testing is executed in the context of wider SDK code and that SDK code applies the validation functions 'later'. So ignore what I said above, sorry!

So, your test shows that at some point in the code we that we _are_ testing will error if the value isn't parsable as a duration ([done here](https://github.com/hashicorp/terraform-provider-google/blob/c70549f5990d90a88e52ec8ee2ac980b8c98d86b/google/provider/provider.go#L1651)) but it's not actually able to test the effects of any validation functions set in the SDK schema.

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/provider/provider_internal_test.go:700]: Could you please also add a test case covering a valid value of request_timeout being added? I think after that we'll have a complete set of tests for this field.

**Review by SarahFrench — COMMENTED** [review]
I think this is the last suggest I have before we get this merged, thanks for all your work on this :shipit:

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/provider/provider_internal_test.go:897]: If you check [this output](https://github.com/GoogleCloudPlatform/magic-modules/actions/runs/5870779051/job/15918512238) from checks running unit tests from this PR it complains about lacking authentication:

```
    --- FAIL: TestProvider_providerConfigure_batching/if_batch_is_configured_with_only_enable_batching (0.03s)
        provider_internal_test.go:952: unexpected error(s): diag.Diagnostics{diag.Diagnostic{Severity:0, Summary:"Attempted to load application default credentials since neither `credentials` nor `access_token` was set in the provider block.  No credentials loaded. To use your gcloud credentials, run 'gcloud auth application-default login'.  Original error: google: could not find default credentials. See https://cloud.google.com/docs/authentication/external/set-up-adc for more information", Detail:"", AttributePath:cty.Path(nil)}}
```

a way to solve this is to supply these fake creds in the `ConfigValues` values for test cases:

```diff
			ConfigValues: map[string]interface{}{
+   			        "credentials":  transport_tpg.TestFakeCredentialsPath,
				"batching": []interface{}{
					map[string]interface{}{
						"enable_batching": true,
					},
				},
			},
```

**Review by SarahFrench — COMMENTED** [review]
Sorry, looks like there's issues with the tests due to how the code managing auth errors out when no application default credentials are present when neither `credentials` or `auth_token` is set. See other comment for suggested fix and link to failing unit test output

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (SarahFrench) [reply mmv1/third_party/terraform/provider/provider_internal_test.go:897]: This is something I've done in equivalent tests elsewhere, e.g. see: https://github.com/GoogleCloudPlatform/magic-modules/pull/7723/files#diff-89e3dcb3dbfe5488060c297a3f4dd633f603f25ca824385b926cd9bec5980d39

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/provider/provider_internal_test.go:703]:
```suggestion
			ConfigValues: map[string]interface{}{
				"request_timeout": "10s",
				"credentials": transport_tpg.TestFakeCredentialsPath,
			},
```

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/provider/provider_internal_test.go:710]:
```suggestion
			ConfigValues: map[string]interface{}{
				"request_timeout": "timeout",
				"credentials": transport_tpg.TestFakeCredentialsPath,
			},
```

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/provider/provider_internal_test.go:726]:
```suggestion
			ConfigValues: map[string]interface{}{
				"request_timeout": "",
				"credentials": transport_tpg.TestFakeCredentialsPath,
			},
```

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/provider/provider_internal_test.go:799]:
```suggestion
			ConfigValues: map[string]interface{}{
				"request_reason": "request test",
				"credentials": transport_tpg.TestFakeCredentialsPath,
			},
```

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/provider/provider_internal_test.go:790]:
```suggestion
			ConfigValues: map[string]interface{}{
				// request_reason unset
				"credentials": transport_tpg.TestFakeCredentialsPath,
			},
			EnvVariables: map[string]string{
```

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/provider/provider_internal_test.go:807]:
```suggestion
			ConfigValues: map[string]interface{}{
				// request_reason unset
				"credentials": transport_tpg.TestFakeCredentialsPath,
			},
			ExpectedConfigValue: "",
```

**Review by SarahFrench — COMMENTED** [review]
Just added a few suggested commits showing how the above recommendation is needed for all tests, including those where the field you're testing isn't set. It isn't ideal, but it's how the auth is set up currently.

**Comment by SarahFrench** [comment]
@BBBmau could you please run `go fmt` on `mmv1/third_party/terraform/provider/provider_internal_test.go`? There's a check complaining about formatting

File: mmv1/third_party/terraform/provider/provider_internal_test.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/provider/provider_internal_test.go:872]: Please add `"credentials": transport_tpg.TestFakeCredentialsPath,` to all the test cases in this test

**Comment by SarahFrench** [comment]
@BBBmau - left two more comments. Could you please build to provider locally and [use `make test` from the root of the generated provider](https://github.com/hashicorp/terraform-provider-google/blob/5f0f5b806c14ead3fba3ac1d8659fc90b042a8cd/GNUmakefile#L11) to make sure the tests in this PR are passing

**Comment by SarahFrench** [comment]
☝️ That failure is because of the 'part 2' PR where I addressed the circular dependency in packages- I was intending on getting this PR done and then updating these tests to include those changes in a follow-up PR. Didn't realise that the automation on PRs would pull in newer changes on main 🤦

**Comment by SarahFrench** [comment]
@BBBmau could you please rebase this PR using the latest version of main and also update these things:

- In the test code, change `providerConfigure` to `provider.ProviderConfigure`
- Similarly, change `p := Provider()` to `p := provider.Provider()`

**Comment by SarahFrench** [comment]
Screenshot 2023-08-30 at 17 04 56

🤦  - I didn't realise that me making a branch with the same name as the branch in your fork of the repo would result in this

**Comment by SarahFrench** [comment]
👆 Addressed a merge conflict that matched the merge conflict we resolved when pairing yesterday - related to removing TestHandleSDKDefaults_RequestReason

**Comment by SarahFrench** [comment]
To address the changes not present in the PR since the last force push, I'm going to try force pushing again to get some fixes in the PR 🤞 . Fighting 🔥 with 🔥

**Review by SarahFrench — APPROVED** [review]
We made it!

celebration-will-smith

**Pull request merged**
